File: DSA/Stack/Code/PostfixToInfix.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

// Postfix to Infix in C++
bool isOperator(char c) {
    return (c == '+' || c == '-' || c == '*' || c == '/');
}

string postfixToInfix(string postfix) {
    stack<string> s;
    
    for (char c : postfix) {
        if (isOperator(c)) {
            string op2 = s.top(); s.pop();
            string op1 = s.top(); s.pop();
            
            string temp = "(" + op1 + c + op2 + ")";
            s.push(temp);
        } else {
            string temp = "";
            temp += c;
            s.push(temp);
        }
    }
    
    return s.top();
}

int main() {
    string postfix = "ab+cd*-";
    
    cout << "=== Postfix to Infix ===\n\n";
    cout << "Postfix Expression: " << postfix << "\n";
    
    string infix = postfixToInfix(postfix);
    
    cout << "Infix Expression: " << infix << "\n\n";
    cout << "Complexity: O(n)\n";
    
    return 0;
}

File: DSA/Stack/Code/PostfixToInfix.java
// Postfix to Infix in Java

import java.util.Stack;

public class PostfixToInfix {
    static boolean isOperator(char c) {
        return (c == '+' || c == '-' || c == '*' || c == '/');
    }
    
    static String postfixToInfix(String postfix) {
        Stack<String> stack = new Stack<>();
        
        for (char c : postfix.toCharArray()) {
            if (isOperator(c)) {
                String op2 = stack.pop();
                String op1 = stack.pop();
                
                String temp = "(" + op1 + c + op2 + ")";
                stack.push(temp);
            } else {
                stack.push(String.valueOf(c));
            }
        }
        
        return stack.pop();
    }
    
    public static void main(String[] args) {
        String postfix = "ab+cd*-";
        
        System.out.println("=== Postfix to Infix ===\n");
        System.out.println("Postfix Expression: " + postfix);
        
        String infix = postfixToInfix(postfix);
        
        System.out.println("Infix Expression: " + infix + "\n");
        System.out.println("Complexity: O(n)");
    }
}
